### PR TITLE
 added support for wolfssl with new bjam option

### PR DESCRIPTION
**Pull request by aldenml**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -54,7 +54,7 @@ set(libtorrent_include_files
 	enum_net
 	error
 	error_code
-	extensions
+	extensions.hpp
```
Comment (arvidn) [inline CMakeLists.txt]: is this right?

File: docs/manual.rst [hunk 2]
```diff
@@ -923,7 +923,7 @@ libtorrent's point of view, it doesn't matter what it is. libtorrent only makes
 sure the peer certificates are signed by the correct root certificate.
 
 One way to create the certificates is to use the ``CA.sh`` script that comes
-with openssl, like thisi (don't forget to enter a common Name for the
+with openssl, like these (don't forget to enter a common Name for the
```
Comment (arvidn) [inline docs/manual.rst]: I just landed a patch in RC_1_2 fixing these typos. this will conflict with that

File: src/peer_connection.cpp [hunk 3]
```diff
@@ -84,10 +84,6 @@ POSSIBILITY OF SUCH DAMAGE.
 #include <set>
 #endif
 
-#ifdef TORRENT_USE_OPENSSL
-#include <openssl/rand.h>
-#endif
-
```
Comment (arvidn) [inline src/peer_connection.cpp]: was this just unused in here?

**Comment by arvidn** [comment]
I'm thinking that there's a simpler way of doing this, perhaps some other way to define the macros to make them a bit more orthogonal.

Also, I don't think this actually solves your problem with WolfSSL all the way. You didn't address setting the SNI to the hostname for HTTPS connections

File: CMakeLists.txt (hunk 1 above)
Comment (aldenml) [reply CMakeLists.txt]: yes, it's something that have been bothering me for a long time. CLion (and `cmake`) pick the folder `include/extensions` as project element, this way only `extensions.hpp` is picked up.

File: docs/manual.rst (hunk 2 above)
Comment (aldenml) [reply docs/manual.rst]: will revert this typo fix then

File: src/peer_connection.cpp (hunk 3 above)
Comment (aldenml) [reply src/peer_connection.cpp]: yes, unusued :)

File: CMakeLists.txt (hunk 1 above)
Comment (arvidn) [reply CMakeLists.txt]: interesting. The purpose of this list of header files is only for creating a package, right?
I just realized there's no logic that adds the ".hpp" suffix to these names. is that implied by cmake?
when building a package with cmake, does it still include the `extensions` directory? I imagine it would need to be included too

Comment (aldenml) [reply CMakeLists.txt]: Yes, `cmake` performs all kind of extensions guessing until it finds (or not) the file. The files under `extensions` are still included with the logic associated to `libtorrent_extensions_include_files`.

File: .travis.yml [hunk 4]
```diff
@@ -23,6 +23,7 @@ matrix:
     - env: variant=test_debug tests=1 toolset=gcc sanitizer=on fuzzers=1
     - env: variant=test_debug tests=1 mmap=off toolset=gcc
     - env: variant=test_debug sim=1 crypto=openssl toolset=gcc sanitizer=on
+    - env: variant=test_debug tests=1 crypto=openssl-https toolset=gcc sanitizer=on
```
Comment (aldenml) [inline .travis.yml]: btw, I have plans to reuse this job for wolfSSL

File: CMakeLists.txt (hunk 1 above)
Comment (arvidn) [reply CMakeLists.txt]: it seems reasonable then to add the `.hpp` extension to all files here, what do you think?

Comment (aldenml) [reply CMakeLists.txt]: hmm... maybe in the future :)

Comment (arvidn) [reply CMakeLists.txt]: I think it would make sense to do this as a separate patch. there's nothing in this PR that depends on this fix, is there?

**Comment by arvidn** [comment]
have you verified that the SNI is set on outgoing HTTPS connections?
currently those calls are conditioned by having a new enough version of openssl, so I suspect they are just compiled out using WolfSSL

**Comment by arvidn** [comment]
I think you have to add "https" to the dictionary: https://travis-ci.org/arvidn/libtorrent/jobs/618948960#L764

File: CMakeLists.txt (hunk 1 above)
Comment (aldenml) [reply CMakeLists.txt]: no, removing this `.hpp` from this PR

**Comment by aldenml** [comment]
I'm attaching images of my IDE debug stepping into SNI code
Screen Shot 2019-11-30 at 2 37 31 PM
and (see `ret = 1`)
Screen Shot 2019-11-30 at 2 39 38 PM

What do you think about the new `TORRENT_USE_OPENSSL_HTTPS_OR_SSL`?

**Comment by aldenml** [comment]
@arvidn do you have an idea of what's happening with https://travis-ci.org/arvidn/libtorrent/jobs/619124602?utm_medium=notification&utm_source=github_status

**Comment by arvidn** [comment]
some of those tests are flaky. at some point I should try to make the more reliable

**Comment by arvidn** [comment]
`TORRENT_USE_OPENSSL_HTTPS_OR_SSL` seems a bit long, and yet not very clear what it means.
What would you think of instead separating out the support for bittorrent over SSL as a separate build configuration that requires SSL? (there could even be a friendly error message if SSL wasn't enabled when trying to enable torrents over SSL).

Something like `TORRENT_BITTORRENT_SSL` or `TORRENT_SSL_PEERS`. That way, we could leave the existing `TORRENT_USE_OPENSSL` as is (I think), to mean to include the basic support for SSL for trackers and web seeds.

Another alternative would be to make it mandatory to build with SSL support, and remove the `TORRENT_USE_OPENSSL` define entirely.

**Comment by aldenml** [comment]
Hi @arvidn, what do you think now with `TORRENT_SSL_PEERS` and `openssl-no-peers`? I know `openssl-no-peers` is not the most beautiful option name, but I fear to break a lot of builds setups and also complicate the build options. Btw, I expect most people to just use `openssl` and not bother with other options.

**Comment by aldenml** [comment]
@arvidn, before fixing the conflict and trigger another CI, what do you think about now that we have a way to separate the Torrents over SSL code,  keeping `openssl` for everything, and moving the `wolfssl` to this PR without the Torrents over SSL?

**Comment by arvidn** [comment]
yes, I think this is a good approach. Would you mind adding `#error` if `TORRENT_SSL_PEERS` is enabled but the openSSL version is too old to support SNI? maybe with a friendly message suggesting this new build configuration `openssl-no-peers`.

Also, if `TORRENT_SSL_PEERS` is set, but `TORRENT_USE_OPENSSL` is *not*, that should also trigger a `#error` with some friendly message.

**Comment by arvidn** [comment]
I would also be OK with making it mandatory to build with a new enough OpenSSL version where SNI is supported.

**Comment by aldenml** [comment]
btw, and sorry if I wasn't clear enough, I want to remove `openssl-no-peers`, and bring `wolfssl`, so it will be only:
`openssl` - everything
`wolfssl` - only https
and the `#error` conditions of your comments. Thoughts?

**Comment by arvidn** [comment]
ah, I see. So, you're thinking a separate define, next to `TORRENT_USE_OPENSSL` for WolfSSL?
so some of the `#ifdef TORRENT_USE_OPENSSL` would also check for your new define?

**Comment by arvidn** [comment]
oh, maybe I misunderstood again. You want to keep the current defines/macros, but you want to rename just the build feature?

**Comment by aldenml** [comment]
correct :), just trying to keep the changes to the minimum, but you tell me

**Comment by arvidn** [comment]
that's fine

**Comment by aldenml** [comment]
rebased, added the wolfSSL here, and added the two #error messages, waiting for CI to complete

**Comment by arvidn** [comment]
it looks like there's something in the random number generator in Wolf causing the current code to fail:

https://travis-ci.org/arvidn/libtorrent/jobs/621878109#L3916

Maybe there's some missing initialization or something.

**Comment by aldenml** [comment]
Fixed the random number generator, there is a subtle API incompatibility, I documented that in the code.

**Comment by arvidn** [comment]
> Now, the arm build is failing consistently, and I can't see the reason for it (maybe the new boost?)

Interestingly, the first test passes. I wonder if it leaves the container in some state that causes the next test to fail.

https://travis-ci.org/arvidn/libtorrent/jobs/622050005#L771

**Comment by aldenml** [comment]
I can confirm that using boost 1.68 builds without problem. Now this is a completely unrelated problem to the wolfSSL integration :( I will take a look at that

**Comment by aldenml** [comment]
travis is green!, appveyor should be fine, waiting for your review

File: .travis.yml [hunk 5]
```diff
@@ -341,5 +352,7 @@ script:
   - cd $TRAVIS_BUILD_DIR/test;
   - 'if [[ "$arch" == "arm" ]];
     then
-      ${B2} arm-tests warnings-as-errors=on warnings=all crypto=$crypto variant=test_arm $toolset target-os=linux link=static testing.launcher="sudo cp -R bin rootfs/; sudo chroot rootfs";
+      ${B2} arm-tests warnings-as-errors=on warnings=all crypto=$crypto variant=test_arm $toolset target-os=linux link=static testing.execute=off;
+      sudo cp -R bin rootfs/;
+      ${B2} arm-tests warnings-as-errors=on warnings=all crypto=$crypto variant=test_arm $toolset target-os=linux link=static testing.launcher="sudo chroot rootfs";
```
Comment (arvidn) [inline .travis.yml]: now that building and running tests are separated, you could even add a `travis_retry` here. (not important though, and I think only deterministic tests are part of the ARM tests)

File: include/libtorrent/aux_/openssl.hpp [hunk 6]
```diff
@@ -101,6 +105,6 @@ inline GENERAL_NAME* openssl_general_name_value(GENERAL_NAMES* gens, int i)
 
 #include "libtorrent/aux_/disable_warnings_pop.hpp"
 
-#endif // TORRENT_USE_OPENSSL
+#endif // TORRENT_SSL_PEERS
```
Comment (arvidn) [inline include/libtorrent/aux_/openssl.hpp]: This comment doesn't look right. I would expect the corresponding `#if` to also have changed in this file

File: src/session_impl.cpp [hunk 7]
```diff
@@ -543,6 +544,8 @@ void apply_deprecated_dht_settings(settings_pack& sett, bdecode_node const& s)
 #ifdef TORRENT_USE_OPENSSL
 		error_code ec;
 		m_ssl_ctx.set_verify_mode(boost::asio::ssl::context::verify_none, ec);
+#endif
+#ifdef TORRENT_SSL_PEERS
 #if OPENSSL_VERSION_NUMBER >= 0x90812f
```
Comment (arvidn) [inline src/session_impl.cpp]: this test can be removed now, right?

Comment (arvidn) [reply src/session_impl.cpp]: and any similar tests, except for the one you added for the `#error`

File: src/socket_type.cpp [hunk 8]
```diff
@@ -279,7 +279,7 @@ namespace aux {
 		TORRENT_UNUSED(holder);
 		error_code e;
 		s.close(e);
-#endif // TORRENT_USE_OPENSSL
+#endif // TORRENT_SSL_PEERS
```
Comment (arvidn) [inline src/socket_type.cpp]: this comment changed, but the corresponding `#if` didn't

File: test/test_session.cpp [hunk 9]
```diff
@@ -333,7 +333,7 @@ TORRENT_TEST(save_state_fingerprint)
 auto const count_dht_inits = [](session& ses)
 {
 	int count = 0;
-	int num = 120; // this number is adjusted per version, an estimate
+	int num = 200; // this number is adjusted per version, an estimate
```
Comment (arvidn) [inline test/test_session.cpp]: this seems unrelated

**Comment by arvidn** [comment]
lgtm. please squash these. However, I think the independent changes should go into separate commits. specifically removing the blank lines at the end of the tests should be a separate commit, and also the independent removal of unused openssl includes as well as the unused `placeholders::_1` using declaration.

File: .travis.yml (hunk 5 above)
Comment (aldenml) [reply .travis.yml]: yes, I will add it

File: src/socket_type.cpp (hunk 8 above)
Comment (aldenml) [reply src/socket_type.cpp]: your are right

File: src/session_impl.cpp (hunk 7 above)
Comment (aldenml) [reply src/session_impl.cpp]: yes, I will remove them

File: include/libtorrent/aux_/openssl.hpp (hunk 6 above)
Comment (aldenml) [reply include/libtorrent/aux_/openssl.hpp]: you are right

File: test/test_session.cpp (hunk 9 above)
Comment (aldenml) [reply test/test_session.cpp]: I was having problems in my env that required this change, will revert it

**Comment by aldenml** [comment]
good, actually, I will create separate PRs for a lot of changes here to strip this one from the non essentials ones.

**Comment by aldenml** [comment]
hi @arvidn, can you restart the travis job? thanks

File: docs/manual.rst [hunk 10]
```diff
@@ -886,7 +886,7 @@ This setting is only taken into account when the normal listen socket is opened
 socket). To not listen on an SSL socket at all, set ``ssl_listen`` to 0.
 
 This feature is only available if libtorrent is build with openssl support
-(``TORRENT_USE_OPENSSL``) and requires at least OpenSSL version 1.0, since it
+(``TORRENT_SSL_PEERS``) and requires at least OpenSSL version 1.0, since it
```
Comment (arvidn) [inline docs/manual.rst]: it seems like this should probably mention both defines, or the text should say something like "built with SSL torrent support"

File: include/libtorrent/config.hpp [hunk 11]
```diff
@@ -116,7 +116,7 @@ POSSIBILITY OF SUCH DAMAGE.
 // on OSX, use the built-in common crypto for built-in
 # if !defined TORRENT_USE_LIBCRYPTO && !defined TORRENT_USE_LIBGCRYPT
 #  define TORRENT_USE_COMMONCRYPTO 1
-# endif // TORRENT_USE_OPENSSL
+# endif
```
Comment (arvidn) [inline include/libtorrent/config.hpp]: was this comment incorrect?

**Comment by arvidn** [comment]
one thing that makes me a little bit nervous is that the headers included when building against wolfSSL have the same name as when linking against openssl. It seems like begging for including the wrong headers.

I imagine if you build against wolfSSL on a system that has openssl "installed", it just won't work. Is that right?

File: docs/manual.rst (hunk 10 above)
Comment (aldenml) [reply docs/manual.rst]: will change

File: include/libtorrent/config.hpp (hunk 11 above)
Comment (aldenml) [reply include/libtorrent/config.hpp]: I think it's incorrect, the if is not with that macro

**Comment by aldenml** [comment]
agree with the fact that this is the perfect recipe for build problems, I looked at this issue from the beginning, but found that the problem begins in `asio`, if you look at https://github.com/boostorg/asio/blob/develop/include/boost/asio/ssl/detail/openssl_types.hpp there is nothing a client library can do (to the best of my knowledge).

I guess this just a "common problem" with C/C++ :)

**Comment by aldenml** [comment]
@arvidn, I don't want to take CI bandwidth for a change in `manual.srt`, do you mind if I do that in another PR?

**Comment by aldenml** [comment]
one travis job restart when you have a chance...thanks

**Comment by arvidn** [comment]
are you sure `test_url_seed` passes? it's failed 6 times in a row now

**Comment by aldenml** [comment]
what can I say :) (green now), but I noticed more random errors lately
